xarray/plot/utils.py

```diff
@@ -98,6 +98,12 @@ def _build_discrete_cmap(cmap, levels, extend, filled):
     # copy the old cmap name, for easier testing
     new_cmap.name = getattr(cmap, "name", cmap)
 
+    # copy colors to use for bad, under, and over values in case they have been set to
+    # non-default values
+    new_cmap._rgba_bad = getattr(cmap, "_rgba_bad", new_cmap._rgba_bad)
```

Comment:
can you raise an issue on the matplotlib repo and ask if there is a better way of doing this?

Reply:
See https://github.com/mwaskom/seaborn/pull/1948

Reply:
The implementation in mwaskom/seaborn#1948 of the same work-around that we are adding here looks nicer and more robust than the way I have implemented it. Will update to follow the same pattern. Not sure if the issue with under/over colors needing to change if they were not explicitly set by the user applies to this case, but in any case I think it is better to handle it in case we ever need to use the same fix somewhere else.

Reply:
Done in de6ffeb and 47a6fc0